### PR TITLE
[DateTimePicker]: Added validation for start/end time

### LINKED ISSUE #3134: [DateTimePicker] User should not be allowed to select a start date after end date
### What package is this for?

- [x] React
- [ ] Angular

### Describe the bug

You can select a start time that's later than the end time. 

### To Reproduce

1. Go to https://carbon-addons-iot-react.com/?path=/story/1-watson-iot-datetime-picker--default
2. Click on DateTimePicker
3. Click on Custom Range
4. Click on the Absolute radio button
5. Change the start date to be after the end date

image

> Please create a reduced test case in CodeSandbox:
> https://codesandbox.io/s/github/carbon-design-system/carbon-addons-iot-react/tree/next/packages/react/examples/codesandbox

### Expected behavior

Apply button should be disabled until start time and end time for absolute time range is valid

### Environment/versions:

- OS: MacOS
- Browser: Chrome
- `carbon-addons-iot-react` version: 2.147.0-next.79

### Specific timeline issues / requests

Do you want this work within a specific time period? Is it related to an
upcoming release?

_NB: The core contributors will try to work with your timeline, but it's not
guaranteed. The earlier you make a request in advance of a desired delivery
date, the better!_

### PR DESCRIPTION
**Pull request by eosineye1**
Closes #3134
**Summary**

- When setting a custom time range for DateTimePicker one can select a start time that's later than the end time.

**Change List (commits, features, bugs, etc)**

- Added validation to prevent a start time after the end time to be selected. An error message is displayed.

**Acceptance Test (how to verify the PR)**

- Try to select a start time after the end time.
- Try to select an end time before the start time.

**Regression Test (how to make sure this PR doesn't break old functionality)**

- Run unit tests 

**Things to look for during review**

- [ ] Make sure all references to `iot` or `bx` class prefix is using the prefix variable
- [ ] (React) All major areas have a `data-testid` attribute. New test ids should have test written to ensure they are not changed or removed.
- [ ] UI should be checked in RTL mode to ensure the proper handling of layout and text.
- [ ] All strings should be translatable.
- [ ] The code should pass a11y scans (The storybook a11y knob should show no violations). New components should have a11y test written.
- [ ] Unit test should be written and should have a coverage of 90% or higher in all areas.
- [ ] All components should be passing visual regression test. For new styles or components either a visual regression test should be written for all permutations or the base image updated.
- [ ] Changes or new components should either write new or update existing documentation.
- [ ] PR should link and close out an existing issue

### CODE REVIEW

**Review by bjornalm — CHANGES_REQUESTED** [review]
Nice work and thanks for contributing!

Code looks good. Could you also write a unit test for this and create an issue in our  carbon-addons-iot-react repo before I approve it?

**Comment by eosineye1** [comment]
@bjornalm Removed some code from old tests as the apply button is disabled for absolute range initially when start and end time are 00:00

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx [hunk 1]
```diff
@@ -229,10 +229,6 @@ describe('DateTimePicker', () => {
     expect(wrapper.find(`.${iotPrefix}--date-time-picker__field`).first().text()).toEqual(
       '2020-04-01 13:34 to 2020-04-06 11:49'
     );
-
-    wrapper.find(`.${iotPrefix}--date-time-picker__menu-btn-apply`).first().simulate('click');
-    jest.runAllTimers();
-    expect(dateTimePickerProps.onApply).toHaveBeenCalled();
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:235]: Why was this removed? The portion above is clicking the increment button setting a valid date and time range, so apply should be enabled and the callback fired.

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx [hunk 2]
```diff
@@ -685,6 +680,40 @@ describe('DateTimePicker', () => {
...
+
+    // Change start time to 02:00 - start time is after end time
+    fireEvent.change(startTime, { target: { value: '02:00' } });
+    expect(applyBytton).toBeDisabled();
+
+    // Change end time to 02:00 - start time is after end time
+    fireEvent.change(endTime, { target: { value: '00:00' } });
+    expect(applyBytton).toBeDisabled();
+  });
+
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:716]: Can you add a little more testing that changes the date and confirms situations where the start time before an end time is valid? ie. start: `02:00` and end: `01:00` is valid if startDate is `2021/10/01` and endDate is `2021/11/01`

**Review by kevinsperrine — CHANGES_REQUESTED** [review]
Thanks for the PR. The code looks good so far. Just a few more minor changes around testing different time ranges and it'll be ready.

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx [hunk 3]
```diff
@@ -494,7 +490,6 @@ describe('DateTimePicker', () => {
     expect(screen.queryByText(i18nDefault.endTimeLabel)).not.toBeInTheDocument();
     // click apply
     fireEvent.click(screen.getByText(i18nTest.applyBtnLabel));
-    expect(screen.getAllByTitle(new RegExp(`.*${i18nTest.toLabel}.*`))[0]).toBeInTheDocument();
```
[the remark this answers is not in the document]
Comment (eosineye1) [reply packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:497]: Added the test back and took your comments into consideration and updated the test.

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx (hunk 1 above)
Comment (eosineye1) [reply packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:235]: Re-added the test to reflect changed made.

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx (hunk 2 above)
Comment (eosineye1) [reply packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:757]: The date wouldn't matter because the date picker will always set the start date as the first date selected and set the last date as the second selected date. So the test mentioned is already covered in the current test cases.

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 4]
```diff
@@ -787,13 +787,35 @@ const DateTimePicker = ({
...
   };
 
+  // Validate that the start time if less than the end time
+  const validateStartEndTime = (startTime, endTime) => {
+    const timefrom = new Date();
+    let temp = startTime.split(':');
+    timefrom.setHours(Number(temp[0]));
+    timefrom.setMinutes(Number(temp[1]));
+
+    const timeto = new Date();
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:797]: You'll need to pass the start/end dates from the absolute range picker into this method, too. By comparing only against the current date with `new Date()` you're showing any `startTime` that's after `endTime` as invalid, but that's not always true. If I have chosen `11/1/2021` as my start date, and `11/3/2021` as my end date. I can have whatever startTime/endTime I want and all of them are valid.

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx (hunk 2 above)
Comment (kevinsperrine) [reply packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:757]: Thanks, @eosineye1, but unfortunately, this isn't correct. The DateTimePicker can be passed a selected date range, so these tests aren't covering the validation correctly. See this [existing story in your preview](https://deploy-preview-3112--carbon-addons-iot-react.netlify.app/?path=/story/1-watson-iot-datetime-picker--selected-absolute). It marks the startTime as invalid, but it _is_ a valid time.

**Review by kevinsperrine — CHANGES_REQUESTED** [review]

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx (hunk 4 above)
Comment (eosineye1) [reply packages/react/src/components/DateTimePicker/DateTimePicker.jsx]: Thanks for letting me know to include the start and end date. I've taken that into consideration and refactored the code.

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 5]
```diff
@@ -787,13 +797,54 @@ const DateTimePicker = ({
...
+      timefrom.setHours(Number(temp[0]));
+      timefrom.setMinutes(Number(temp[1]));
+
+      // Include end time in end date
+      const timeto = new Date(absoluteValues.end);
+      temp = endTime.split(':');
+      timeto.setHours(Number(temp[0]));
+      timeto.setMinutes(Number(temp[1]));
+
+      console.log(timefrom, timeto);
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:816]: Let's remove the console.log.

```suggestion

```

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx [hunk 6]
```diff
@@ -685,6 +694,86 @@ describe('DateTimePicker', () => {
...
+    const defaultValues = {
+      timeRangeKind: PICKER_KINDS.ABSOLUTE,
+      timeRangeValue: {
+        startDate: '2020-04-01',
+        startTime: '12:34',
+        endDate: '2020-04-01',
+        endTime: '11:49',
+      },
+    };
+    const wrapper = mount(<DateTimePicker {...dateTimePickerProps} defaultValue={defaultValues} />);
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:707]: Looking great. Thanks, @eosineye1. I should have mentioned this yesterday, but didn't think of it. Can you switch these tests over to using testing-library instead of enzyme. We're working on replacing all the existing enzyme tests and write all new tests in testing library.

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx [hunk 7]
```diff
@@ -685,6 +694,86 @@ describe('DateTimePicker', () => {
...
+    jest.runAllTimers();
+    expect(wrapper.find(`.${iotPrefix}--date-time-picker__field`)).toHaveLength(1);
+    expect(wrapper.find(`.${iotPrefix}--date-time-picker__field`).first().text()).toEqual(
+      '2020-04-01 12:34 to 2020-04-01 11:49'
+    );
+
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:713]: We can make these test more concrete by checking for the existence of the "Invalid time format." message, so here, for example, we could check:
```jsx
expect(screen.getAllByText('Invalid time format.')).toHaveLength(2);
expect(screen.getByRole('button', {name: 'Apply'})).toBeDisabled();
```

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx [hunk 8]
```diff
@@ -685,6 +694,86 @@ describe('DateTimePicker', () => {
...
+    expect(wrapper.find(`.${iotPrefix}--date-time-picker__field`).first().text()).toEqual(
+      '2020-04-01 13:34 to 2020-04-01 12:49'
+    );
+
+    wrapper.find(`.${iotPrefix}--time-picker__controls--btn.up-icon`).at(1).simulate('click');
+    jest.runAllTimers();
+    expect(wrapper.find(`.${iotPrefix}--date-time-picker__field`).first().text()).toEqual(
+      '2020-04-01 13:34 to 2020-04-01 13:49'
+    );
+
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:731]: And then here the opposite of above:

```jsx
expect(screen.queryByText('Invalid time format.')).toBeNull();
expect(screen.getByRole('button', {name: 'Apply'})).not.toBeDisabled();

File: packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx [hunk 9]
```diff
@@ -685,6 +694,86 @@ describe('DateTimePicker', () => {
...
+        endTime: '11:49',
+      },
+    };
+    const wrapper = mount(<DateTimePicker {...dateTimePickerProps} defaultValue={defaultValues} />);
+    jest.runAllTimers();
+    expect(wrapper.find(`.${iotPrefix}--date-time-picker__field`)).toHaveLength(1);
+    expect(wrapper.find(`.${iotPrefix}--date-time-picker__field`).first().text()).toEqual(
+      '2020-04-01 12:34 to 2020-04-06 11:49'
+    );
+
```
Comment (kevinsperrine) [inline packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx:753]: same checks as comments in test above to check the invalid message does not appear with valid times and button is enabled (and vice-versa).

**Review by kevinsperrine — CHANGES_REQUESTED** [review]
That's awesome, @eosineye1. Looks great while testing it in storybook now. Just a few comments to make the jest tests more concrete and this is ready to go. Thanks for this addition, It's a good improvement on the component.

**Comment by eosineye1** [comment]
Hi @kevinsperrine, I am having issues adding default dates and times to the DateTimePicker using the testing library. I am currently going through the docs, but can't find resources about passing in default values. If you know a way to do so using the testing library could you point me in the right direction?

**Comment by kevinsperrine** [comment]
> 

Sure thing. You can pass them in the `deafultValue` prop when render is called. You can look at some of the other tests. For example: 

https://github.com/carbon-design-system/carbon-addons-iot-react/blob/1f3e2513117d4d98fdc811e14a285a5f50f0b110/packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx#L531

https://github.com/carbon-design-system/carbon-addons-iot-react/blob/1f3e2513117d4d98fdc811e14a285a5f50f0b110/packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx#L552

Or an example of dynamically typing in the time boxes: 

https://github.com/carbon-design-system/carbon-addons-iot-react/blob/1f3e2513117d4d98fdc811e14a285a5f50f0b110/packages/react/src/components/DateTimePicker/DateTimePicker.test.jsx#L669

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 10]
```diff
@@ -787,13 +797,52 @@ const DateTimePicker = ({
...
   // on change functions that trigger a absolute value update
   const onAbsoluteStartTimeChange = (pickerValue, evt, meta) => {
     setAbsoluteStartTimeInvalid(meta.invalid);
+    const invalidStartEndTime = invalidAbsoluteDate(
+      pickerValue,
+      absoluteValue.endTime,
+      absoluteValue
+    );
+    setAbsoluteStartTimeInvalid(meta.invalid || invalidStartEndTime);
+    setAbsoluteEndTimeInvalid(absoluteEndTimeInvalid && invalidStartEndTime);
```
Comment (bjornalm) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:832]: This will cause a bug if you:
1. Select an absolute date range 
2. Input an invalid number like `9999` in the "End time" input
3. Enter a valid number in the "Start time" input.
4. The "End time" input is still invalid but will no longer be shown as invalid because of the `absoluteEndTimeInvalid && invalidStartEndTime` where `invalidStartEndTime` is false

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 11]
```diff
@@ -787,13 +797,52 @@ const DateTimePicker = ({
...
   const onAbsoluteEndTimeChange = (pickerValue, evt, meta) => {
     setAbsoluteEndTimeInvalid(meta.invalid);
+    const invalidStartEndTime = invalidAbsoluteDate(
+      absoluteValue.startTime,
+      pickerValue,
+      absoluteValue
+    );
+    setAbsoluteEndTimeInvalid(meta.invalid || invalidStartEndTime);
+
+    setAbsoluteStartTimeInvalid(absoluteStartTimeInvalid && invalidStartEndTime);
```
Comment (bjornalm) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:844]: Same bug here

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 12]
```diff
@@ -787,13 +797,52 @@ const DateTimePicker = ({
...
+      pickerValue,
+      absoluteValue.endTime,
+      absoluteValue
+    );
+    setAbsoluteStartTimeInvalid(meta.invalid || invalidStartEndTime);
+    setAbsoluteEndTimeInvalid(absoluteEndTimeInvalid && invalidStartEndTime);
     changeAbsolutePropertyValue('startTime', pickerValue);
   };
   const onAbsoluteEndTimeChange = (pickerValue, evt, meta) => {
     setAbsoluteEndTimeInvalid(meta.invalid);
```
Comment (bjornalm) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:836]: Can be removed

**Review by bjornalm — CHANGES_REQUESTED** [review]

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx (hunk 10 above)
Comment (eosineye1) [reply packages/react/src/components/DateTimePicker/DateTimePicker.jsx]: Logic has been updated to account for this and to validate each TimrSpinner individually. Also, a test has been added to test this

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 13]
```diff
@@ -787,13 +795,60 @@ const DateTimePicker = ({
     setAbsoluteValue(newAbsolute);
   };
 
+  const isValidDate = (d, time) => {
```
Comment (bjornalm) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:798]: We are getting there :-) This check will however incorrectly return true for any time value following `nn:nn`, e.g. the value `99:99`. If you instead of splitting the time before and and pass in the full time string you could use the same regex used in the TimePickerSpinner.jsx:

```
const isValid24HoursRegex = /^([01][0-9]|2[0-3]):([0-5][0-9])$/;
const isValidTime = isValid24HoursRegex.test(time);
```

Ideally this would not be duplicated but in a separate file and imported int both components..

**Review by bjornalm — CHANGES_REQUESTED** [review]

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx (hunk 13 above)
Comment (eosineye1) [reply packages/react/src/components/DateTimePicker/DateTimePicker.jsx]: Utilizing the regrex in the isValidDate function not. I was not sure where to put the regrex and was not sure where to put a new file if I made one for me to import it for both components, so I copied it into the DateTimePicker component file.

**Review by bjornalm — APPROVED** [review]
Nice job on this one!
There are still some edge cases not covered by the tests but I think we can merge this now since the functionality is ok. Just commit the suggested spelling fix.

**Review by davidicus — APPROVED** [review]
Should be good once Kevin approves

**Review by kevinsperrine — COMMENTED** [review]
I swear, I thought I did this yesterday, but I'm not seeing the comment now. Nice work, @eosineye1. Just a few linting errors and this is ready to merge!

```carbon-addons-iot-react: /home/runner/work/carbon-addons-iot-react/carbon-addons-iot-react/packages/react/src/components/DateTimePicker/DateTimePicker.jsx
carbon-addons-iot-react:   598:7   error  'invalidEndDate' was used before it was defined    no-use-before-define
carbon-addons-iot-react:   601:7   error  'invalidStartDate' was used before it was defined  no-use-before-define
carbon-addons-iot-react:   800:37  error  Unexpected use of 'isNaN'                          no-restricted-globals```

**Review by kevinsperrine — APPROVED** [review]

File: packages/react/src/components/DateTimePicker/DateTimePicker.jsx [hunk 14]
```diff
@@ -701,13 +701,15 @@ const DateTimePicker = ({
...
         if (!absolute.hasOwnProperty('start')) {
-          absolute.start = dayjs(absolute.startDate).valueOf();
+          absolute.start = dayjs(`${absolute.startDate} ${absolute.startTime}`).valueOf();
         }
         if (!absolute.hasOwnProperty('end')) {
-          absolute.end = dayjs(absolute.endDate).valueOf();
+          absolute.end = dayjs(`${absolute.endDate} ${absolute.endTime}`).valueOf();
         }
         absolute.startDate = dayjs(absolute.start).format('MM/DD/YYYY');
+        absolute.startTime = dayjs(absolute.start).format('HH:mm');
```
Comment (stuckless) [inline packages/react/src/components/DateTimePicker/DateTimePicker.jsx:710]: Seem wrong to me that we are hard coding the date/time formats.  We should be looking at a way to externize this.  I don't think it is simple enough to expose a "pattern" since, what is the format of the pattern?  we can't have different components using different kinds of patterns... some using moment, some using dayjs, etc.

Comment (stuckless) [reply packages/react/src/components/DateTimePicker/DateTimePicker.jsx:710]: I think PAL needs to have a standardize set of of API around globalization and that object should be made available to components that need to parse/format data, get access to masks, etc.

In graphite when we use "formats" we use named formats like "short_date", "long_date", "date_fully_month", etc, we translate those into the appropriate masks based on user locale.

Comment (kevinsperrine) [reply packages/react/src/components/DateTimePicker/DateTimePicker.jsx:710]: Thanks, Sean. We've talked about this internally and have started creating a document defining data formatting, i18n, data loading, etc--all the main items we've talked about over the last few months. We're moving that direction. This particular component was written 2 years ago, and this PR is to add time validation to the component as it already exists. The refactors based on new technical design will come later.

**Review by davidicus — APPROVED** [review]
